### PR TITLE
State recovery

### PR DESCRIPTION
**Pull request by Lederstrumpf**

### CODE REVIEW

**Comment by TheCharlatan** [comment]
Rebased.

**Comment by TheCharlatan** [comment]
This `KeyManager` serialization is not working, so this seems to need a patch in -core in order to continue the work.

**Comment by h4sh3d** [comment]
> This `KeyManager` serialization is not working, so this seems to need a patch in -core in order to continue the work.

Done in https://github.com/farcaster-project/farcaster-core/pull/239, update `farcaster_core` in lock to https://github.com/farcaster-project/farcaster-core/commit/0f7f8c58c5d8fa88be108c9f6457ac975cb0ab86

**Comment by TheCharlatan** [comment]
I'm out of ideas on how to fix the hanging unit tests. Any ideas @h4sh3d ?

**Comment by h4sh3d** [comment]
> I'm out of ideas on how to fix the hanging unit tests. Any ideas @h4sh3d ?

I have to dig deeper, but I'll take time to see what's going on. Maybe reorganizing the tests at the same time will help to catch edge cases.

**Comment by TheCharlatan** [comment]
Ok, fixed it ^^

File: src/checkpointd/runtime.rs [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+impl From<Vec<u8>> for CheckpointKey {
+    fn from(raw_key: Vec<u8>) -> Self {
+        CheckpointKey {
+            swap_id: SwapId(raw_key[0..32].try_into().unwrap()),
+            service_id: raw_key[32..].to_vec().into(),
+        }
+    }
+}
+
+struct CheckpointGetSet(lmdb::Environment);
```
[the remark this answers is not in the document]
Comment (zkao) [reply src/checkpointd/runtime.rs:257]: additionally, i find the multipart message should not be handled above internet2, but at the level of zmq, that is by exposing a socket, and let zmq deal with the multipart messages

[the remark this answers is not in the document]
Comment (zkao) [reply src/checkpointd/runtime.rs:257]: except for the wire protocol messages (msgs exchanged btw peers)

**Review by zkao — COMMENTED** [review]
Generally very good PR, left many comments, hopefully comprehensible enough

Summary:
- zmq handles multipart messages, and ideally that should happen at that lower layer
- most of the checkpointd is just an interface to a database

so basically its like an interface to a db through a socket, i'd say

I guess this PR can be merged if addressing changes seem too dramatic and blocking, but first I'd like us, especially @TheCharlatan, to have some careful though over this one

File: src/checkpointd/runtime.rs [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+                if self.pending_checkpoint_chunks.contains_key(&checksum) {
+                    let chunks = self
+                        .pending_checkpoint_chunks
+                        .get_mut(&checksum)
+                        .expect("checked with contains_key");
+                    chunks.insert(CheckpointChunk {
+                        msg_index,
+                        serialized_state_chunk,
+                    });
+                } else {
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/checkpointd/runtime.rs:144]: I don't understand what you are asking here. Either it adds to the chunks if a chunk with the key already exists, or it creates a new entry for that chunk if a key does not exist yet.
I also don't understand what you mean with "zmq does it simpler". Do you mean I should use the msg length instead of its index? What if two messages are the same size?

File: src/checkpointd/runtime.rs [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+
+            Request::RetrieveAllCheckpointInfo => {
+                let _pairs = self.checkpoints.get_all_key_value_pairs();
+            }
+
+            Request::DeleteCheckpoint(swap_id) => {
+                self.checkpoints
+                    .delete_state(CheckpointKey {
+                        swap_id,
+                        service_id: ServiceId::Wallet,
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/checkpointd/runtime.rs:209]: I did not want every service to clean up its checkpoint data independently. Rather, it is the responsibility of `farcasterd` now to issue this request, which seemed more in tune with how we handle the other post-swap cleanup routines.

File: src/checkpointd/runtime.rs [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+        pending_checkpoint_chunks: map![],
+    };
+
+    Service::run(config, runtime, false)
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+struct CheckpointChunk {
+    pub msg_index: usize,
+    pub serialized_state_chunk: Vec<u8>,
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/checkpointd/runtime.rs:49]: Using a bit to signal more messages arriving works if the messages are guaranteed to arrive in order. While this is a guarantee that holds here even with our complex architecture in practice, I don't wholly trust it, since some things could go wrong during routing, and think it is more defensive to give each of the chunks an index and an expected total end size.

File: src/checkpointd/runtime.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/checkpointd/runtime.rs]: Rename of checkpointd and CHeckpointGetSet done in 89ebf9f1c0343a90ef4425fdfda4772b57da796e

File: src/checkpointd/runtime.rs [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+        identity: ServiceId::Checkpoint,
+        checkpoints: CheckpointGetSet::new(data_dir).unwrap(),
+        pending_checkpoint_chunks: map![],
+    };
+
+    Service::run(config, runtime, false)
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+struct CheckpointChunk {
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/checkpointd/runtime.rs]: I was reluctant to implement a generic message chunker, because I do not want this to be a permanent solution. The permanent solution is fixing internet2 . That said, I'd be happy to instead of chunking the checkpoint data, chunk the enitre request and rename it accordingly. Then the generic names line more up with the more generic nature of our Request enum. Is this what you are asking for?

**Comment by zkao** [comment]
> Yeah, it totally should. The chunking is really just a silly workaround of this internet-2 limitation. Instead of spending a lot of time on making the maximum message size in internet2 configurable, passing review and then waiting for a release, I went ahead and implemented this a as a temporary solution that can be removed easily and without architectural baggage once the this has been fixed in internet2.

But we still can do it below internet2 layer and strict encoding, can't we? By exposing a socket pair to the runtimes that need to communicate with databased. Then databased is not built over internet2, only over zmq sockets. Why wouldn't this work?

File: src/checkpointd/runtime.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (zkao) [reply src/checkpointd/runtime.rs]: you'd have to add extra headers to the message, i'd guess:

- length
- checksum
- serviceid? (depending on the socket type)

[the remark this answers is not in the document]
Comment (zkao) [reply src/checkpointd/runtime.rs]: But then send it as a raw zmq message to the databased runtime, that knows how to decode those messages correctly

**Comment by TheCharlatan** [comment]
> But we still can do it below internet2 layer and strict encoding, can't we? By exposing a socket pair to the runtimes that need to communicate with databased. Then databased is not built over internet2, only over zmq sockets. Why wouldn't this work?
> I'm not so sure that would belong to internet2

Sure, but I do not want to sacrifice our architecture for the sake of an artificially introduced limitation. The way I see it, we can't use any of the internet 2 dependencies for this, so we would have to hand-roll an entire zmq server-router-client architecture and make it intrerop with internet 2 in order to get this done and then bin it again once the limitation is removed. Not worth the effort in my eyes compared to what I wrote here within the better half of an afternoon.

File: src/checkpointd/runtime.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/checkpointd/runtime.rs]: That's exactly what is in the message right now. Length is provided by strict encoding, and serviceid and checksum are part of the message struct.

[the remark this answers is not in the document]
Comment (zkao) [reply src/checkpointd/runtime.rs]: but that you're doing over the service bus, what im saying is do it over zmq directly

**Comment by zkao** [comment]
> The way I see it, we can't use any of the internet 2 dependencies for this, so we would have to hand-roll an entire zmq server-router-client architecture and make it intrerop with internet 2 in order to get this done and then bin it again once the limitation is removed

what is it that is missing from internet2 that you believe belongs there? 

im not totally sure this belongs to internet2, that's the thing, it feels its something lower level

**Comment by zkao** [comment]
> I can't think of a way to inject lower level zmq things into the existing internet2 sutff in order to get large messages working.

REQ-ROUTER-DEALER-REP,  i think does it, but i will study the possibility of farcasterd doing it. the client right now is REQ-ROUTER-ROUTER-REP works as well https://github.com/farcaster-project/farcaster-node/blob/890723cb27f0db14a0d871664528a899f45f1f42/src/rpc/client.rs#L54

**Comment by zkao** [comment]
https://github.com/Internet2-WG/rust-internet2/blob/d3025d56608e7dd2f525c2262edbc7264b80d325/src/transport/zeromq.rs#L470-L501

**Comment by TheCharlatan** [comment]
> here internet2 exposes the low level stuff, you used it in the syncer, you don't think we can use it here?
> 
> https://github.com/Internet2-WG/rust-internet2/blob/d3025d56608e7dd2f525c2262edbc7264b80d325/src/transport/zeromq.rs#L470-L501

I know about this, however the last time I tried to manipulate the sender and receiver,  I ended up re-building most of the internet 2 library because the relevant fields that would have allowed lower level manipulation were private. In that way it is possible, it's just going to be very intrusive since you need to re-do the plumbing for most of our microservices.

File: src/walletd/runtime.rs [hunk 6]
```diff
@@ -1382,6 +1544,56 @@ impl Runtime {
...
+pub fn checkpoint_state(
+    endpoints: &mut Endpoints,
+    swap_id: SwapId,
+    state: request::CheckpointState,
+) -> Result<(), Error> {
+    let mut serialized_state = vec![];
+    let size = state.strict_encode(&mut serialized_state).unwrap();
+
+    // if the size exceeds a boundary, send a multi-part message
+    let max_chunk_size = internet2::transport::MAX_FRAME_SIZE - 1024;
```
Comment (zkao) [inline src/walletd/runtime.rs:1556]: I find the the client daemons (here wallet) to have to multipart messages not the correct api, but i guess if this code was in databased, it would feel more natural, its just an import for the client daemon

**Comment by zkao** [comment]
> relevant fields that would have allowed lower level manipulation were private

what fields are needed?

I don't like the api with the clients having to multipart msgs, that is it. if this can be abstracted such that clients dont notice this, i guess, i would be fine, even if doing above internet2, ideally not there

i'd say we should contribute to internet2 if its missing there, even if we hack around it for now

in any case, i'm ok with this getting merged, but logging those issues

function to access to zmq socket, what else is needed?
https://github.com/Internet2-WG/rust-internet2/blob/d3025d56608e7dd2f525c2262edbc7264b80d325/src/transport/zeromq.rs#L349-L351

**Comment by TheCharlatan** [comment]
> what fields are needed?

On my last attempt I had to re-define the connection struct, since it controls how the sending and receiving is performed. It's field are private and give you no control on the frame handling. For the existing manipulations in our code base, we did not have to change the frame handling, so it is much easier to implement.

> i'd say we should contribute to internet2 if its missing there, even if we hack around it for now

I want to talk to Maxim and see if he's open to just drop the length restrictions there, or at least open to making them configurable.

**Comment by TheCharlatan** [comment]
Large messages will be allowed after https://github.com/Internet2-WG/rust-internet2/pull/26#event-6820409869 makes it into a release.
